src/DessserQCheck.ml

```diff
@@ -173,61 +173,45 @@ let shrink_mac_type mt =
   loop to_simplest
 
 let rec shrink_value_type =
+  let open Iter in
   let vt_of_mn mn = mn.T.vtyp in
-  let shrink_fields mns make_typ f =
-    Array.iter (fun (_, mn) -> shrink_maybe_nullable mn (f % vt_of_mn)) mns ;
-    let shrink_mns =
-      let shrink (fn, mn) =
-        Iter.map (fun mn -> fn, mn) (shrink_maybe_nullable mn) in
-      Shrink.filter (fun mns -> Array.length mns > 1)
-        (Shrink.array ~shrink) mns |>
-      Iter.map make_typ in
-    shrink_mns f in
+  let shrink_fields f =
+    (Shrink.pair Shrink.nil shrink_maybe_nullable) f in
   function
-  | T.Unknown | T.Ext _ ->
-      Iter.empty
-  | T.Base Unit ->
-      Iter.empty
-  | T.Base mt ->
-      (fun f ->
-        shrink_mac_type mt (fun mt -> f (T.Base mt)))
-  | T.Usr _ ->
-      Iter.empty
+  | T.Unknown | T.Ext _ -> empty
+  | T.Base Unit -> empty
+  | T.Base mt -> shrink_mac_type mt >|= T.base
+  | T.Usr _ -> empty
   | T.Vec (dim, mn) ->
-      (fun f ->
-        shrink_maybe_nullable mn (fun mn ->
-          f mn.vtyp ;
-          f (T.Vec (dim, mn))))
+      let smn = shrink_maybe_nullable mn in
+      (smn >|= T.vec dim) <+> (smn >|= vt_of_mn)
   | T.Lst mn ->
-      (fun f ->
-        shrink_maybe_nullable mn (fun mn ->
-          f (T.Lst mn) ;
-          f mn.vtyp))
+      let smn = shrink_maybe_nullable mn in
+      (smn >|= T.lst) <+> (smn >|= vt_of_mn)
   | T.Set (st, mn) ->
-      (fun f ->
-        shrink_maybe_nullable mn (fun mn ->
-          f (T.Set (st, mn)) ;
-          f mn.vtyp))
+      let smn = shrink_maybe_nullable mn in
+      (smn >|= T.set st) <+> (smn >|= vt_of_mn)
   | T.Tup mns ->
-      (fun f ->
-        Array.iter (fun mn -> shrink_maybe_nullable mn (f % vt_of_mn)) mns ;
-        let shrink_mns =
-          Shrink.filter (fun mns -> Array.length mns > 1)
-            (Shrink.array ~shrink:shrink_maybe_nullable) mns |>
-          Iter.map (fun mns -> T.Tup mns) in
-        shrink_mns f)
+      let smns = Shrink.array ~shrink:shrink_maybe_nullable mns in
+      filter (fun mn -> Array.length mn > 1) smns >|= T.tup
+       <+>
+      (of_array mns >|= vt_of_mn)
   | T.Rec mns ->
-      shrink_fields mns (fun mns -> T.Rec mns)
+      let smns = Shrink.array ~shrink:shrink_fields mns in
+      filter (fun mn -> Array.length mn > 1) smns >|= T.record
+       <+>
+      (of_array mns >|= vt_of_mn % snd)
   | T.Sum mns ->
-      shrink_fields mns (fun mns -> T.Sum mns)
+      let smns = Shrink.array ~shrink:shrink_fields mns in
+      filter (fun mn -> Array.length mn > 1) smns >|= T.sum
+       <+>
+      (of_array mns >|= vt_of_mn % snd)
   | T.Map (k, v) ->
-      (fun f ->
-        shrink_maybe_nullable k (f % vt_of_mn) ;
-        shrink_maybe_nullable v (f % vt_of_mn) ;
-        let shrink_kv =
-          (Shrink.pair shrink_maybe_nullable shrink_maybe_nullable) (k, v) |>
-          Iter.map (fun (k, v) -> T.Map (k, v)) in
-        shrink_kv f)
+      let sk = shrink_maybe_nullable k in
+      let sv = shrink_maybe_nullable v in
```

Comment:
That's good that you moved all those calls to shrink_maybe_nullable out of the function body and into the closure; Given it often end up calling shrink_max_type,  that probably explains the speed improvement.

Reply:
Also, I'm not certain but given a quick look at qcheck code it seems that Shrink.pair is not the same as Iter.pair. I would even go as far as calling Shrink.iter buggy.